src/Space.ts

```diff
@@ -1,11 +1,116 @@
+class MemberUpdateEvent extends Event {
+  constructor(public members: SpaceMember[]) {
+    super('memberUpdate', {});
+  }
 }
 
+const createSpaceMemberFromPresenceMember = (m: Types.PresenceMessage): SpaceMember => ({
+  clientId: m.clientId as string,
+  isConnected: true,
+  data: JSON.parse(m.data as string),
+});
+
+class Space extends EventTarget {
+  private members: SpaceMember[] = [];
+  private channelName: string;
+  private channel: Types.RealtimeChannelPromise;
+
+  eventTarget: EventTarget;
+
+  constructor(private name: string, private client: Types.RealtimePromise, private options?: SpaceOptions) {
+    super();
+    this.setChannel(this.name);
+  }
+
+  private setChannel(rootName) {
+    // The channel name prefix here should be unique to avoid conflicts with non-space channels
+    this.channelName = `_ably_space_${rootName}`;
+    this.channel = this.client.channels.get(this.channelName);
+  }
+
+  private createMember(clientId: string, isConnected: boolean, data: { [key: string]: any }) {
+    this.members.push({ clientId, isConnected, data });
+  }
+
+  private syncMembers() {
+    this.channel.presence.get().then((presenceMessages) => {
+      this.members = presenceMessages.filter((m) => m.clientId).map(createSpaceMemberFromPresenceMember);
+    });
+  }
+
+  private subscribeToPresenceEvents() {
+    this.channel.presence.subscribe('enter', (message: Types.PresenceMessage) => {
+      this.updateMemberState(message.clientId, true, JSON.parse(message.data as string));
+    });
+
+    this.channel.presence.subscribe('leave', (message: Types.PresenceMessage) => {
+      this.updateMemberState(message.clientId, false);
+    });
+
+    this.channel.presence.subscribe('update', (message: Types.PresenceMessage) => {
+      this.updateMemberState(message.clientId, true, JSON.parse(message.data as string));
+    });
+  }
+
+  private updateMemberState(clientId: string | undefined, isConnected: boolean, data?: { [key: string]: any }) {
+    const implicitClientId = clientId ?? this.client.auth.clientId;
+
+    if (!implicitClientId) {
+      return;
+    }
+
+    const member = this.members.find((m) => m.clientId === clientId);
+
+    if (!member) {
+      this.createMember(implicitClientId, isConnected, data || {});
+    } else {
+      member.isConnected = isConnected;
+      if (data) {
+        // Member data is completely overridden, except lastEventTimestamp which is updated
+        member.data = {
+          ...data,
+          lastEventTimestamp: new Date(),
+        };
+      }
+    }
+
+    const memberUpdateEvent = new MemberUpdateEvent(this.members);
+    this.dispatchEvent(memberUpdateEvent);
+  }
+
+  enter(data: unknown) {
+    if (!data || typeof data !== 'object') {
+      return;
+    }
+
+    const clientId = this.client.auth.clientId || undefined;
+    const presence = this.channel.presence;
+
+    // TODO: Discuss if we actually want change this behaviour in contrast to presence (enter becomes an update)
```

Comment:
I think `enter()` should be idempotent: it's shouldn't reject if you're already entered. (Or of the client thinks you're already entered.) It's also not necessary to convert it to an `update()` - realtime will do this if it sees the member already present.

Reply:
I'll create a small ticket for this.

Reply:
thanks, @paddybyers. We've added this (and left it as TODO) because of what knock-on effect it might have on UI frameworks. The main issue here is that updates should not trigger unnecessary renders for UI. So—data updates, goes to a UI via a `subscribe`, but ideally, if state has not actually changed no renders are triggers (because they can be expensive). The idea here is to prevent this. I wanted to test this in practice, with React and see where we can provide a solution for developers, so this is not an issue.